### PR TITLE
Implement [[msvc::no_unique_address]]

### PR DESCRIPTION
**Pull request by amykhuang**
This implements the [[msvc::no_unique_address]] attribute.

There is not ABI compatibility in this patch because the attribute is relatively new and there's still some uncertainty in the MSVC version.

Bug: https://github.com/llvm/llvm-project/issues/49358

Also see https://reviews.llvm.org/D157762.

### CODE REVIEW

File: clang/lib/AST/Decl.cpp
Comment (rnk) [inline clang/lib/AST/Decl.cpp:4518]: I think you've addressed this FIXME. Perhaps you can early return true if !MS ABI, and then return a result based on the any_of check.

File: clang/lib/AST/RecordLayoutBuilder.cpp
Comment (rnk) [inline clang/lib/AST/RecordLayoutBuilder.cpp:3198]: If we need to track DataSize, should we create an abstraction for updating both members?

Comment (amykhuang) [reply clang/lib/AST/RecordLayoutBuilder.cpp:3198]: I realized I can reduce the places that DataSize is set, and I think it's more readable this way?

**Comment by cor3ntin** [comment]
We should consider whether we want to support __msvc_no_unique_address__ or similar
https://github.com/llvm/llvm-project/issues/61196

**Comment by philnik777** [comment]
> We should consider whether we want to support `__msvc_no_unique_address__` or similar as an alternative spelling #61196

I think `[[__msvc__::__no_unique_address__]]` would be better. This is how the clang and gnu attributes are handled too.

File: clang/lib/AST/RecordLayoutBuilder.cpp
Comment (amykhuang) [reply clang/lib/AST/RecordLayoutBuilder.cpp:2984]: Actually going to remove this altogether. Makes it less consistent with MSVC's behavior but MSVC appears to avoid overlapping fields with bases only when no_unique_address is used, which seems weird.

File: clang/lib/AST/RecordLayoutBuilder.cpp
Comment (cjdb) [inline clang/lib/AST/RecordLayoutBuilder.cpp:3142]: I'd prefer it if we didn't have cascading assignment please.

**Review by cjdb — CHANGES_REQUESTED** [review]

**Review by AaronBallman — COMMENTED** [review]
Added @erichkeane as attributes code owner, and @efriedma-quic and @rjmccall as codegen code owners.

The attributes changes look good to me. The codegen changes seem reasonable, but are a bit outside my area of expertise for me to feel comfortable signing off on. The only thing I did spot was that this should come with a change to the release notes so users know about the new functionality and its stability guarantees.

**Comment by efriedma-quic** [comment]
When you say you want a second opinion on CodeGen changes, do you mean RecordLayoutBuilder.cpp?  I don't see any non-whitespace changes to clang/lib/CodeGen/ .

File: clang/lib/AST/RecordLayoutBuilder.cpp
Comment (efriedma-quic) [inline clang/lib/AST/RecordLayoutBuilder.cpp:3010]: Are primary virtual bases actually an thing in the MSVC ABI?  It seems like PrimaryVirtualBase and PrimaryVirtualBaseInfo aren't actually used anywhere.

File: clang/include/clang/Basic/Attr.td
Comment (erichkeane) [inline clang/include/clang/Basic/Attr.td:1804]:
```suggestion
  let Accessors = [Accessor<"isStandard", [CXX11<"", "no_unique_address", 201803>]>,
```

File: clang/lib/AST/Decl.cpp
Comment (erichkeane) [inline clang/lib/AST/Decl.cpp:4508]: Does any differentiation need to be made between the spellings on 4489?  Additionally, can you elaborate on the comment here?  How can we look at the fields of `CXXRD` if it is empty?

**Review by erichkeane — COMMENTED** [review]
The `RecordLayoutBuilder` is not something I have any knowledge about, but I shared two comments.

**Comment by AaronBallman** [comment]
> When you say you want a second opinion on CodeGen changes, do you mean RecordLayoutBuilder.cpp? I don't see any non-whitespace changes to clang/lib/CodeGen/ .

Sorry for the confusion, yes, I was mostly worried about RecordLayoutBuilder.cpp and whether you spotted any concerns with the changes.

File: clang/lib/AST/RecordLayoutBuilder.cpp
Comment (amykhuang) [reply clang/lib/AST/RecordLayoutBuilder.cpp]: Removed cascading assignments

File: clang/lib/AST/RecordLayoutBuilder.cpp
Comment (amykhuang) [reply clang/lib/AST/RecordLayoutBuilder.cpp]: I removed this function for the time being, but yes, that part shouldn't be there.

**Review by cjdb — APPROVED** [review]
Thanks! LGTM from me.

File: clang/lib/AST/Decl.cpp
Comment (amykhuang) [reply clang/lib/AST/Decl.cpp]: As far as I can tell the cases above still apply to MSVC so I think on 4489 it's correct to account for both spellings. Not sure what you mean about the comment? The existing comment above the `isEmpty` line describes what that checks for

File: clang/lib/AST/RecordLayoutBuilder.cpp
Comment (mstorsjo) [inline clang/lib/AST/RecordLayoutBuilder.cpp:2877]: Nit: After adding and removing things here, the total result of the PR ends up in adding new spurious empty lines here (and right above, and below on line 2949 and 2967).

File: clang/lib/CodeGen/CGRecordLayoutBuilder.cpp
Comment (mstorsjo) [inline clang/lib/CodeGen/CGRecordLayoutBuilder.cpp:746]: Nit: The total PR sums up to adding a spurious empty line here, while this file otherwise no longer has any functional changes.

**Comment by cor3ntin** [comment]
> > We should consider whether we want to support `__msvc_no_unique_address__` or similar as an alternative spelling #61196
> 
> I think `[[__msvc__::__no_unique_address__]]` would be better. This is how the clang and gnu attributes are handled too. I'm also fine with handling this in a follow-up patch.

Yes, apparently, that is what tablegen is doing automatically. We might just need to add a test

**Review by cor3ntin — COMMENTED** [review]
There are some whitespace-only that should be reverted + 1 small nitpick.
And I have a question about `__has_cpp_attribute`

File: clang/test/Preprocessor/has_attribute.cpp
Comment (AaronBallman) [reply clang/test/Preprocessor/has_attribute.cpp:58]: Correct, this appears to be a bug (possibly with the way we're generating code from ClangAttrEmitter.cpp); this should be `0` on `WINDOWS`. I think this is happening because we removed `TargetSpecificAttr` from the attribute definition -- if I'm correct, we may require some deeper changes to express target-specific *spellings* of an attribute instead of making the attribute as a whole target-specific. CC @erichkeane as there's possibly going to be some design questions there.

File: clang/test/SemaCXX/cxx2a-no-unique-address.cpp
Comment (AaronBallman) [inline clang/test/SemaCXX/cxx2a-no-unique-address.cpp:2]: This also appears to be a bug; given the Windows triple, this should still be giving "unknown attribute ignored" warnings.

File: clang/test/Preprocessor/has_attribute.cpp
Comment (AaronBallman) [reply clang/test/Preprocessor/has_attribute.cpp:58]: OOoh, actually, here's an idea that *might* work without needing to do tablegen surgery:
```
def NoUniqueAddress : InheritableAttr {
  let Subjects = SubjectList<[NonBitField], ErrorDiag>;
  let Documentation = [NoUniqueAddressDocs];
}

def MSNoUniqueAddress : NoUniqueAddress, TargetSpecificAttr<TargetMicrosoftCXXABI> {
  let Spellings = [CXX11<"msvc", "no_unique_address", 201803>];
}

def ItaniumNoUniqueAddress : NoUniqueAddress, TargetSpecificAttr<TargetItaniumCXXABI> {
  let Spellings = [CXX11<"", "no_unique_address", 201803>];
}
```
We'd have to do `let AdditionalMembers = ` in the base class to manually implement what is done automatically with `let Accessors =`, but this might work. Worth noting: totally untested idea, so it could also explode horribly.

File: clang/lib/AST/Decl.cpp
Comment (erichkeane) [inline clang/lib/AST/Decl.cpp:4516]: My point here was, the isEmpty test above means we already know that CXXRD has no non-static data members of class type: https://clang.llvm.org/doxygen/classclang_1_1CXXRecordDecl.html#ac570e2716d2cb020851575aa4e5fafe0

So this test doesn't seem to do anything.

Comment (amykhuang) [reply clang/lib/AST/Decl.cpp:4516]: Oh, I see. The comment above the isEmpty check specifies "-- has subobjects of nonzero size or bit-fields of nonzero length", so I was assuming that isEmpty is taking the zero size stuff into account. I found it mentioned in this comment as well https://github.com/llvm/llvm-project/blob/004a7cea705d63441307a5e0429b6786d49042c2/clang/include/clang/AST/CXXRecordDeclDefinitionBits.def#L49. 

Also, in the last commit I added tests to verify that this check is working.

File: clang/test/Preprocessor/has_attribute.cpp
Comment (amykhuang) [reply clang/test/Preprocessor/has_attribute.cpp]: Tried this and the inheritance part wasn't quite working (think that the NoUniqueAddress should be a class, but in that case, not sure how to parse the two attributes). I did a slightly different version where we have three attributes with no inheritance (MS, Itanium, and general) and just parse the MS and Itanium versions as the general version in SemaDeclAttr. It at least passes the tests that were broken before.

**Comment by amykhuang** [comment]
> > We should consider whether we want to support `__msvc_no_unique_address__` or similar as an alternative spelling #61196
> 
> I think `[[__msvc__::__no_unique_address__]]` would be better. This is how the clang and gnu attributes are handled too. I'm also fine with handling this in a follow-up patch.

Sounds good, I can do this in a separate patch

File: clang/lib/AST/RecordLayoutBuilder.cpp
Comment (shafik) [inline clang/lib/AST/RecordLayoutBuilder.cpp:3356]:
```suggestion
      MicrosoftRecordLayoutBuilder Builder(*this, /*EmptySubobjects=*/nullptr);
```

Comment (shafik) [reply clang/lib/AST/RecordLayoutBuilder.cpp:3356]: see [bugprone-argument-comment](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/argument-comment.html)

**Comment by amykhuang** [comment]
Are there any other comments here? I think I've addressed all the existing ones.

**Comment by MaskRay** [comment]
The supposed doc fix (for `doc-clang-html`?) #67195 caused `error: Record `NoUniqueAddress', field `Documentation' exists but does not have a list value`. I have reverted #67195.

**Pull request merged**
